Author: Rene Kita <mail@rkta.de>

Enable re-ordering of buffers in the stack menu

I'm not happy with this implementation.

My original version did not use 'Buffer *prev'. It would check if we are at the
bottom of the stack with

    if (CurrentMenu->select == SelectMenu.nitem - 2) /* Bottom of menu */
       return (MENU_NOTHING);

but then GCC would warn about a potential null-dereference at

    if (cur == Firstbuf)
       Firstbuf = cur->nextBuffer;
    else
       prevBuffer(Firstbuf, cur)->nextBuffer = next;

because prevBuffer could return NULL if there were no previous buffer (which
can't happen because of the previous check).

Given the choice between losing a useful warning or wasting CPU cycles getting
prev even if we don't need it, I decided to do the unnecessary computation
because it's no big deal nowadays. Now I feel dirty like a whore.

---
 menu.c |   74 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 74 insertions(+)
--- a/menu.c
+++ b/menu.c
@@ -242,6 +242,8 @@ static int SelectV = 0;
 static void initSelectMenu(void);
 static void smChBuf(void);
 static int smDelBuf(char c);
+static int smMovDn(char c);
+static int smMovUp(char c);
 
 /* --- SelectMenu (END) --- */
 
@@ -1442,6 +1444,8 @@ initSelectMenu(void)
     SelectMenu.cursorX = Currentbuf->cursorX + Currentbuf->rootX;
     SelectMenu.cursorY = Currentbuf->cursorY + Currentbuf->rootY;
     SelectMenu.keymap['D'] = smDelBuf;
+    SelectMenu.keymap['>'] = smMovUp;
+    SelectMenu.keymap['<'] = smMovDn;
     SelectMenu.item[nitem].type = MENU_NOP;
 }
 
@@ -1506,6 +1510,76 @@ smDelBuf(char c)
     return (MENU_NOTHING);
 }
 
+static int
+smMovUp(char c)
+{
+    Buffer *cur, *prev, *pprev;
+    int i, mselect, x, y;
+
+    if (!CurrentMenu->select) /* Top of menu */
+	return MENU_NOTHING;
+    for (i = 0, cur = Firstbuf; i < CurrentMenu->select;
+	 i++, cur = cur->nextBuffer) ;
+    prev = prevBuffer(Firstbuf, cur);
+    if ((pprev = prevBuffer(Firstbuf, prev)))
+	pprev->nextBuffer = cur;
+    prev->nextBuffer = cur->nextBuffer;
+    cur->nextBuffer = prev;
+    if (prev == Firstbuf)
+	Firstbuf = cur;
+
+    x = CurrentMenu->x;
+    y = CurrentMenu->y;
+    mselect = CurrentMenu->select;
+
+    initSelectMenu();
+    CurrentMenu->x = x;
+    CurrentMenu->y = y;
+    geom_menu(CurrentMenu, 0, 0, 0);
+    CurrentMenu->select = --mselect;
+    displayBuffer(Currentbuf, B_FORCE_REDRAW);
+    draw_all_menu(CurrentMenu);
+    select_menu(CurrentMenu, CurrentMenu->select);
+    return MENU_NOTHING;
+}
+
+static int
+smMovDn(char c)
+{
+    Buffer *cur, *next, *prev;
+    int i, mselect, x, y;
+
+    /* Get selected buffer */
+    for (i = 0, cur = Firstbuf; i < CurrentMenu->select;
+	 i++, cur = cur->nextBuffer) ;
+
+    /* Nothing to do if now previous buffer */
+    if (!(prev = prevBuffer(Firstbuf, cur)))
+	return (MENU_NOTHING);
+
+    next = cur->nextBuffer;
+    if (cur == Firstbuf)
+	Firstbuf = cur->nextBuffer;
+    else
+	prev->nextBuffer = next;
+    cur->nextBuffer = next->nextBuffer;
+    next->nextBuffer = cur;
+
+    x = CurrentMenu->x;
+    y = CurrentMenu->y;
+    mselect = CurrentMenu->select;
+
+    initSelectMenu();
+    CurrentMenu->x = x;
+    CurrentMenu->y = y;
+    geom_menu(CurrentMenu, 0, 0, 0);
+    CurrentMenu->select = ++mselect;
+    displayBuffer(Currentbuf, B_FORCE_REDRAW);
+    draw_all_menu(CurrentMenu);
+    select_menu(CurrentMenu, CurrentMenu->select);
+    return MENU_NOTHING;
+}
+
 /* --- SelectMenu (END) --- */
 
 /* --- SelTabMenu --- */
